Implement game save plumbing and enable for missions

pchote opened the pull request:
This PR implements game saving and loading using order resimulation.

This implements a couple of tricks to speed the process up vs replays:
 * World rendering is disabled, and UI rendering capped at 5 FPS
 * Sync calculations are disabled until the very end

Also note that chat and other immediate orders are not saved (to save confusion when clients change in MP saves).

On my machine I see a 30-40x real time restoration for one of the TD campaign missions.

Another big difference from replays is that traits can store arbitrary data that is returned on load - this is used to restore the viewport and selection groups. In the future we could expand on this if we want to implement a proper save/load using serialization.

Saves are enabled for campaign missions, and compatible with (but disabled for) skirmishes. A followup PR will enable skirmish savegames after implementing `IGameSaveTraitData` on the bot modules to properly restore their state.

This also lays the groundwork for multiplayer saves, with several TODOs noting the remaining work to implement these.

Closes #2743.
Supersedes #6287.

pchote committed:
Add player trait support for IWorldLoaded.

pchote committed:
Add Sound.DisableAllSounds property.

pchote committed:
Fix TD input prompt button height.

pchote committed:
Save and restore viewport position / selection / control groups.

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -810,18 +920,50 @@ public void StartGame()
+			// Enable game saves for singleplayer missions only
+			// TODO: Enable for skirmish once the AI supports state-restoration
+			// TODO: Enable for multiplayer (non-dedicated servers only) once the lobby UI has been created
+			LobbyInfo.GlobalSettings.GameSavesEnabled = !Dedicated && Map.Visibility == MapVisibility.MissionSelector &&
```
Remove the `Map.Visibility == MapVisibility.MissionSelector` here to enable saving in skirmish.

reaperrr commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/GameSaveLoadingLogic.cs:
```diff
@@ -0,0 +1,51 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
```
2019

pchote commented:
Fixed.

reaperrr commented:
Can't say much about the code, other than that it seems to work (loading a skirmish game seemingly doesn't work even without bots, but I assume that's to be expected and will be fixed with the next PR either way).
Saving and loading missions seems to work flawlessly, didn't see any UI regressions either.
Cautious 👍 from me, though ideally I'd like to see @obrakmann, @chrisforbes or @RoosterDragon to check and approve the code before it's merged.
Some more testing from others wouldn't hurt, either.

pchote commented:
The first pushed version had issues with skirmish games, but these should now be fixed (provided you comment out the check that disables them).

matjaeck commented:
Don't know why but I can't load into these [replays.zip](https://github.com/OpenRA/OpenRA/files/3076823/replays.zip)

Unlike the filenames might tell, it's not caused by saving during demolition, this works as expected.

Phrohdoh commented on OpenRA.Game/Network/GameSave.cs:
```diff
@@ -0,0 +1,287 @@
+			// File format:
+			// - List of orders in network frame format
+			// - Metadata start marker
+			//   - Last frame containing orders (int32)
```
Frame number or index?

matjaeck commented:
Aborting loading of a saved game reliably breaks any ingame sounds when I start a mission after it.

Got this exception while trying to load into one of the replays posted above (unfortunately don't know which one), after extracting again from the zip archive.

<details>

```
OpenRA engine version {DEV_VERSION}
Red Alert mod version {DEV_VERSION}
Date: 2019-04-13 23:48:30Z
Operating System: Linux (Unix 4.15.0.47)
Runtime Version: Mono 5.18.1.0 (tarball Fri Mar 15 20:41:32 UTC 2019) CLR 4.0.30319.42000
Exception of type `System.InvalidOperationException`: Out of sync in frame 177.
 Compare syncreport.log with other players.
  at OpenRA.Network.OrderManager.OutOfSync (System.Int32 frame) [0x00037] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Network.OrderManager.CheckSync (System.Byte[] packet) [0x00034] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Network.OrderManager+<>c__DisplayClass40_0.<TickImmediate>b__3 (System.Int32 clientId, System.Byte[] packet) [0x00038] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Network.ReplayConnection.Receive (System.Action`2[T1,T2] packetFn) [0x00052] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Network.OrderManager.TickImmediate () [0x000cb] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Sync+<>c__DisplayClass13_0.<RunUnsynced>b__0 () [0x00000] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Sync.RunUnsynced[T] (System.Boolean checkSyncHash, OpenRA.World world, System.Func`1[TResult] fn) [0x00023] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Sync.RunUnsynced (System.Boolean checkSyncHash, OpenRA.World world, System.Action fn) [0x0000d] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x000f6] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Game.LogicTick () [0x00071] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Game.Loop () [0x000e0] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Game.Run () [0x0003c] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <9c96625fb1da40059cf161732c28f14c>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in <9c96625fb1da40059cf161732c28f14c>:0
``` 

</details>

Might not be related but my hotkey setup (game, unit, unit stances) has partly been reset after starting a mission. The viewport and production hotkeys however have not been reset.

pchote commented:
@matjaeck did you try to reload a save from an earlier version of the PR using the latest version? I believe I rebased, which may well have introduced sync-affecting changes.

matjaeck commented:
These have been saved using your remote branch. But I'll fetch the PR like I would do usually and try to reproduce to make sure.

matjaeck commented:
@pchote I can not guarantee that these are from the the updated branch so better consider the comment as false positive for now. But I can reproduce the mentioned sound issue .

pchote commented:
Fixed the sound issue.

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -105,7 +110,7 @@ void CacheChatLine(Color color, string name, string text)
 		{
 			var immediateOrders = localOrders.Where(o => o.IsImmediate).ToList();
-			if (immediateOrders.Count != 0)
+			if (immediateOrders.Count != 0 && GameSaveLastFrame < NetFrameNumber + FramesAhead)
```
NetFrameNumber + FramesAhead is used in several places, could be made to its own property

pchote replied:
Can you suggest a name for this?

reaperrr commented:
Can confirm that both saving and loading bot-less skirmishes now works.

pchote committed:
Implement game save/load backend.

pchote committed:
Dynamically populate the ingame menu.

pchote commented:
Updated and rebased (so any existing saves will now be invalid).

pchote commented on OpenRA.Game/Network/GameSave.cs:
```diff
@@ -0,0 +1,287 @@
+			Team = client.Team;
+			Slot = client.Slot;
+			Bot = client.Bot;
+			BotControllerClientIndex = client.BotControllerClientIndex;
```
This is probably bogus - we need to remap the client index to a SlotClient index.
Not a problem for mission saves or even skirmish saves, but I expect this will cause problems in the far future of MP saves.

pchote replied:
I removed this dodgy serialization in favor of a documented workaround (see the `// HACK` comment in `DispatchOrders`) that only needs to be resolved if we ever implement MP saves.

reaperrr commented:
Loading skirmishes works and couldn't reproduce that sound issue after last updates.

One thing, though:
In my opinion, the `GameSaveViewportManager` commit should get at least a notification update rule that tells modders to manually add the trait, and ideally refer them to a wiki entry detailing required UI changes, too.
UI changes have by now become the biggest showstopper for easy transitions to newer engine versions, as they (usually) don't get update rules and cannot be caught by the linter.

pchote commented:
Indeed, this will have its own section on the [Update Notes](https://github.com/OpenRA/OpenRAModSDK/wiki/Updating-to-a-new-SDK-or-Engine-version) for the next release.

pchote commented:
Updated and rebased - savegames from previous versions are again invalidated.

This should now work for all skirmish configurations if enabled as commented above, with the caveat that bots will start the reloaded game without any squad configurations or other history, so may act a bit strange until they recover and reassign squads.

reaperrr commented:
> OpenRA.Game/Server/Server.cs:L933: [SA1003] The spacing around the symbol '&&' is invalid.

pchote commented:
I accidentally commited my test change to enable skirmish saves - fixed.

I have also created https://github.com/OpenRA/OpenRAModSDK/wiki/Update-notes:-release-20190314-to-next#game-saves as mentioned above.

pchote committed:
Allow the cursor to be force disabled.

pchote committed:
Implement loading screens.

pchote committed:
Add save/load browser and buttons to the UI.

pchote committed:
Implement "Game Saved" / "Game Loaded" notifications in RA and D2k.

pchote commented:
Updated:
 * Fixed all of @matjaeck's requests.
 * Added a new commit working around an ugly issue with replays ignoring state.
 * Split the cursor hiding into its own commit, as this is now shared with #16414.
 * Moved some fixups that were accidentally in the wrong commits where they belong.

pchote commented:
I suspect I have a better solution for the replay issue - to be investigated.

pchote commented:
Fixed the replay issue properly by sending the trait data before the orders stream and dropped the redundant workaround commit.

matjaeck commented:
Thanks for fixing this here, I've tested it again and everything seems so work so consider my :+1: still valid.

reaperrr reviewed (approved):
Looks good as far as I can tell, SP savegames still work, and I trust @matjaeck's testing more than my own anyway so LGTM.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/799bf42299140fcd0d5652d4e8a49b697c357f33...d969c6a25ea8d2a102eccc5eb044d7e67db46521
